File: sdk/java/src/main/java/run/runtime/sdk/SDKError.java
package run.runtime.sdk;

import java.util.Map;
import java.util.Objects;

public final class SDKError extends RuntimeException {
  private static final long serialVersionUID = 1L;

  private final ErrorCode code;
  private final String stage;
  private final RetryHint retryHint;
  private final boolean retryable;
  private final String source;
  private final String invocationId;
  private final String receiptURA;
  private final transient Map<String, Object> details;

  public SDKError(
      ErrorCode code,
      String stage,
      RetryHint retryHint,
      boolean retryable,
      String message,
      String source,
      String invocationId,
      String receiptURA,
      Map<String, Object> details,
      Throwable cause) {
    super(required(message, "message"), cause);
    this.code = Objects.requireNonNull(code, "code");
    this.stage = required(stage, "stage");
    this.retryHint = Objects.requireNonNullElse(retryHint, RetryHint.NEVER);
    this.retryable = retryable;
    this.source = source == null ? "" : source;
    this.invocationId = invocationId == null ? "" : invocationId;
    this.receiptURA = receiptURA == null ? "" : receiptURA;
    this.details = details == null ? Map.of() : Map.copyOf(details);
  }

  public static SDKError validation(String stage, String message) {
    return new SDKError(
        ErrorCode.INVALID_ARGUMENT,
        stage,
        RetryHint.NEVER,
        false,
        message,
        "",
        "",
        "",
        Map.of(),
        null);
  }

  public static SDKError closed(String stage) {
    return new SDKError(
        ErrorCode.INVALID_HANDLE,
        stage,
        RetryHint.NEVER,
        false,
        stage + " is closed",
        "",
        "",
        "",
        Map.of(),
        null);
  }

  public ErrorCode code() {
    return code;
  }

  public String stage() {
    return stage;
  }

  public RetryHint retryHint() {
    return retryHint;
  }

  public boolean retryable() {
    return retryable;
  }

  public String source() {
    return source;
  }

  public String invocationId() {
    return invocationId;
  }

  public String receiptURA() {
    return receiptURA;
  }

  public Map<String, Object> details() {
    return details;
  }

  public ErrorClass errorClass() {
    return switch (code) {
      case INVALID_ARGUMENT, NULL_POINTER, INVALID_UTF8, INVALID_INVOCATION ->
          ErrorClass.VALIDATION;
      case INVALID_HANDLE -> ErrorClass.HANDLE;
      case NOT_INITIALIZED, ALREADY_INIT -> ErrorClass.LIFECYCLE;
      case RUNTIME_OFFLINE, TRANSPORT -> ErrorClass.AVAILABILITY;
      case PERMISSION_DENIED, HTTP_AUTH_DENIED, CALLER_IDENTITY_UNAVAILABLE ->
          ErrorClass.PERMISSION;
      case ADMISSION_DENIED, SIGNATURE_DENIED, POLICY_DENIED, AUTHORITY_DENIED,
              AUTHORITY_SUBJECT_MISMATCH, EXECUTION_FAILED, ABILITY_FAILED,
              CALLER_SIGNER_UNAVAILABLE, RECEIPT_PROOF_FACTS_MISSING ->
          ErrorClass.ADMISSION;
      case ABILITY_NOT_FOUND, ROUTE_UNAVAILABLE, NOT_FOUND, DESCRIPTOR_NOT_FOUND,
              DESCRIPTOR_OWNER_OFFLINE, DESCRIPTOR_MODE_UNSUPPORTED, DESCRIPTOR_STALE,
              RUNTIME_ROUTE_UNAVAILABLE, PROVIDER_UNAVAILABLE ->
          ErrorClass.ROUTING;
      case TIMEOUT, INVOCATION_TIMEOUT -> ErrorClass.TIMEOUT;
      case CANCELLED, INVOCATION_CANCELLED -> ErrorClass.CANCELLATION;
      case PROTOCOL, PROTOCOL_MISMATCH -> ErrorClass.PROTOCOL;
      case VERSION_MISMATCH, VERSION_INCOMPATIBLE -> ErrorClass.VERSION;
      case CONTROL_ONLY -> ErrorClass.CONTROL;
      case NOT_IMPLEMENTED -> ErrorClass.UNSUPPORTED;
      case TERMINAL_RECEIPT_UNAVAILABLE, GENERIC -> ErrorClass.GENERIC;
    };
  }

  private static String required(String value, String field) {
    if (value == null || value.isBlank()) {
      throw new IllegalArgumentException(field + " is required");
    }
    return value;
  }
}

File: sdk/swift/Sources/RuntimeSDK/SDKError.swift
import Foundation

public enum SDKErrorCode: String, Sendable {
    case invalidArgument = "INVALID_ARGUMENT"
    case invalidHandle = "INVALID_HANDLE"
    case nullPointer = "NULL_POINTER"
    case invalidUTF8 = "INVALID_UTF8"
    case notInitialized = "NOT_INITIALIZED"
    case alreadyInit = "ALREADY_INIT"
    case runtimeOffline = "RUNTIME_OFFLINE"
    case permissionDenied = "PERMISSION_DENIED"
    case admissionDenied = "ADMISSION_DENIED"
    case httpAuthDenied = "HTTP_AUTH_DENIED"
    case signatureDenied = "SIGNATURE_DENIED"
    case policyDenied = "POLICY_DENIED"
    case authorityDenied = "AUTHORITY_DENIED"
    case authoritySubjectMismatch = "AUTHORITY_SUBJECT_MISMATCH"
    case abilityNotFound = "ABILITY_NOT_FOUND"
    case routeUnavailable = "ROUTE_UNAVAILABLE"
    case executionFailed = "EXECUTION_FAILED"
    case timeout = "TIMEOUT"
    case cancelled = "CANCELLED"
    case invalidInvocation = "INVALID_INVOCATION"
    case protocolMismatch = "PROTOCOL_MISMATCH"
    case versionMismatch = "VERSION_MISMATCH"
    case versionIncompatible = "VERSION_INCOMPATIBLE"
    case controlOnly = "CONTROL_ONLY"
    case transport = "TRANSPORT"
    case protocolFailure = "PROTOCOL"
    case notFound = "NOT_FOUND"
    case abilityFailed = "ABILITY_FAILED"
    case notImplemented = "NOT_IMPLEMENTED"
    case generic = "GENERIC"
    case callerIdentityUnavailable = "CALLER_IDENTITY_UNAVAILABLE"
    case callerSignerUnavailable = "CALLER_SIGNER_UNAVAILABLE"
    case descriptorNotFound = "DESCRIPTOR_NOT_FOUND"
    case descriptorOwnerOffline = "DESCRIPTOR_OWNER_OFFLINE"
    case descriptorModeUnsupported = "DESCRIPTOR_MODE_UNSUPPORTED"
    case descriptorStale = "DESCRIPTOR_STALE"
    case runtimeRouteUnavailable = "RUNTIME_ROUTE_UNAVAILABLE"
    case invocationCancelled = "INVOCATION_CANCELLED"
    case invocationTimeout = "INVOCATION_TIMEOUT"
    case terminalReceiptUnavailable = "TERMINAL_RECEIPT_UNAVAILABLE"
    case receiptProofFactsMissing = "RECEIPT_PROOF_FACTS_MISSING"
    case providerUnavailable = "PROVIDER_UNAVAILABLE"
}

public enum SDKErrorClass: String, Sendable {
    case validation
    case handle
    case lifecycle
    case availability
    case permission
    case admission
    case routing
    case timeout
    case cancellation
    case protocolFailure = "protocol"
    case version
    case control
    case unsupported
    case generic
}

public enum RetryHint: String, Sendable {
    case never
    case safe
    case afterBackoff = "after_backoff"
    case unknown
}

public struct SDKError: Error, Sendable, CustomStringConvertible {
    public let code: SDKErrorCode
    public let stage: String
    public let retryHint: RetryHint
    public let retryable: Bool
    public let message: String
    public let source: String
    public let invocationID: String
    public let receiptURA: String
    public let details: [String: String]

    public init(
        code: SDKErrorCode,
        stage: String,
        retryHint: RetryHint = .never,
        retryable: Bool = false,
        message: String,
        source: String = "",
        invocationID: String = "",
        receiptURA: String = "",
        details: [String: String] = [:]
    ) {
        self.code = code
        self.stage = SDKError.required(stage, "stage")
        self.retryHint = retryHint
        self.retryable = retryable
        self.message = SDKError.required(message, "message")
        self.source = source
        self.invocationID = invocationID
        self.receiptURA = receiptURA
        self.details = details
    }

    public static func validation(_ stage: String, _ message: String) -> SDKError {
        SDKError(code: .invalidArgument, stage: stage, message: message)
    }

    public static func closed(_ stage: String) -> SDKError {
        SDKError(code: .invalidHandle, stage: stage, message: "\(stage) is closed")
    }

    public var errorClass: SDKErrorClass {
        switch code {
        case .invalidArgument, .nullPointer, .invalidUTF8, .invalidInvocation:
            return .validation
        case .invalidHandle:
            return .handle
        case .notInitialized, .alreadyInit:
            return .lifecycle
        case .runtimeOffline, .transport:
            return .availability
        case .permissionDenied, .httpAuthDenied, .callerIdentityUnavailable:
            return .permission
        case .admissionDenied, .signatureDenied, .policyDenied, .authorityDenied, .authoritySubjectMismatch, .executionFailed, .abilityFailed, .callerSignerUnavailable, .receiptProofFactsMissing:
            return .admission
        case .abilityNotFound, .routeUnavailable, .notFound, .descriptorNotFound, .descriptorOwnerOffline, .descriptorModeUnsupported, .descriptorStale, .runtimeRouteUnavailable, .providerUnavailable:
            return .routing
        case .timeout, .invocationTimeout:
            return .timeout
        case .cancelled, .invocationCancelled:
            return .cancellation
        case .protocolFailure, .protocolMismatch:
            return .protocolFailure
        case .versionMismatch, .versionIncompatible:
            return .version
        case .controlOnly:
            return .control
        case .notImplemented:
            return .unsupported
        case .terminalReceiptUnavailable, .generic:
            return .generic
        }
    }

    public var description: String {
        "\(code.rawValue): \(message)"
    }

    static func required(_ value: String?, _ field: String) -> String {
        guard let value, !value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return field
        }
        return value
    }
}
